File: easy/q20/q20.java
import java.util.Stack;

class q20 {
    public boolean isValid(String s) {
        char[] ar = s.toCharArray();
        Stack<Character> stack = new Stack<>();
        for (char c : ar) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) {return false;}
                char top = stack.pop();

                if ((c == ')' && top != '(') ||
                    (c == '}' && top != '{') ||
                    (c == ']' && top != '[')) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}

File: easy/q20/q20.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = Vec::new();
        let chars: Vec<char> = s.chars().collect();

        for &c in chars.iter() {
            match c {
                '(' | '{' | '[' => stack.push(c),
                ')' | '}' | ']' => {
                    if let Some(top) = stack.pop() {
                        if (c == ')' && top != '(') || (c == '}' && top != '{') || (c == ']' && top != '[') {
                            return false;
                        }
                    } else {
                        return false;
                    }
                }
                _ => {}
            }
        }

        stack.is_empty()
    }
}
